noir-projects/aztec-nr/aztec/src/macros/notes/mod.nr

```diff
@@ -552,43 +628,123 @@ comptime fn generate_finalization_payload(
     };
 
     // We compute the log length and we concatenate the fields into a single quote.
-    let log_length = fields_list.len();
+    let public_values_length = fields_list.len();
     let fields = fields_list.join(quote {,});
 
     // Now we compute quotes relevant to the multi-scalar multiplication.
-    let (generators_list, scalars_list, args_list, msm_aux_vars) =
+    let (generators_list, _, args_list, msm_aux_vars) =
         generate_multi_scalar_mul(indexed_nullable_fields);
 
+    // We generate scalars_list manually as we need it to refer self.public_values
+    let mut scalars_list: [Quoted] = &[];
+    for i in 0..public_values_length {
+        scalars_list =
+            scalars_list.push_back(quote { std::hash::from_field_unsafe(self.public_values[$i]) });
+    }
+
     let generators = generators_list.join(quote {,});
     let scalars = scalars_list.join(quote {,});
     let args = args_list.join(quote {,});
 
+    // Then we compute values for `encrypt_log(...)` function
+    let setup_log_plaintext_length = indexed_fixed_fields.len() * 32 + 64;
+    // 353 is a fixed ciphertext overhead as defined by `compute_encrypted_note_log`
+    let setup_log_byte_length = 353 + setup_log_plaintext_length;
+    // Each field contains 31 bytes so the length in fields is computed as ceil(setup_log_byte_length / 31)
+    // --> we achieve rouding by adding 30 and then dividing without remainder
+    let setup_log_field_length = (setup_log_byte_length + 30) / 31;
+    let finalization_log_byte_length = setup_log_byte_length + public_values_length * 32;
+
     (
         quote {
         struct $finalization_payload_name {
-            log: [Field; $log_length],
-            note_hash: Field,
+            context: &mut aztec::prelude::PublicContext,
+            hiding_point_slot: Field,
+            setup_log_slot: Field,
+            public_values: [Field; $public_values_length],
         }
 
         impl $finalization_payload_name {
-            fn new(mut self, hiding_point: aztec::protocol_types::point::Point, $args) -> $finalization_payload_name {
+            fn new(mut self, context: &mut aztec::prelude::PublicContext, slot: Field, $args) -> $finalization_payload_name {
+                self.context = context;
+
+                self.hiding_point_slot = slot;
+                self.setup_log_slot = slot + aztec::protocol_types::point::POINT_LENGTH as Field;
+
                 $aux_vars_for_serialization
-                self.log = [$fields];
+                self.public_values = [$fields];
+
+                self
+            }
+
+            fn emit(self) {
+                self.emit_note_hash();
+                self.emit_log();
```

Comment:
I kept the emit_note_hash and emit_log functions exposed on the struct as it a dev might not want to emit logs (e.g. when delivering via some off-chain method).